src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosDbCollectionPhysicalPartitionInfo.cs

```diff
@@ -0,0 +1,166 @@
+        {
+            EnsureArg.IsNotNull(dataStoreConfiguration, nameof(dataStoreConfiguration));
+            EnsureArg.IsNotNull(collectionConfiguration, nameof(collectionConfiguration));
+            EnsureArg.IsNotNull(httpClientFactory, nameof(httpClientFactory));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _dataStoreConfiguration = dataStoreConfiguration;
+            _collectionConfiguration = collectionConfiguration.Get(Constants.CollectionConfigurationName);
+            _httpClientFactory = httpClientFactory;
+            _logger = logger;
+        }
+
+        public int PhysicalPartitionCount { get; private set; }
+
+        public async Task EnsureInitialized()
+        {
+            PhysicalPartitionCount = await GetPhysicalPartitionCount(_backgroundLoopCancellationTokenSource.Token);
+            _backgroundLoopTask = BackgroundLoop(_backgroundLoopCancellationTokenSource.Token);
+        }
+
+        public async Task BackgroundLoop(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                try
+                {
+                    await Task.Delay(TimeSpan.FromMinutes(10), cancellationToken);
+                    PhysicalPartitionCount = await GetPhysicalPartitionCount(cancellationToken);
+                }
+                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+                {
+                    return;
+                }
+                catch (Exception e)
+                {
+                    _logger.LogError("Unable to get physical partition count.", e);
+                }
+            }
+        }
+
+        private async Task<int> GetPhysicalPartitionCount(CancellationToken cancellationToken)
+        {
+            using HttpClient client = _httpClientFactory.CreateClient();
+
+            string host = _dataStoreConfiguration.Host;
+            string key = _dataStoreConfiguration.Key;
+
+            if (string.IsNullOrWhiteSpace(host) && string.IsNullOrWhiteSpace(key))
+            {
+                host = CosmosDbLocalEmulator.Host;
+                key = CosmosDbLocalEmulator.Key;
+            }
+
+            string date = DateTime.UtcNow.ToString("R");
+
+            bool isResourceToken = IsResourceToken(key);
+
+            string authToken = HttpUtility.UrlEncode(
+                isResourceToken
+                    ? key
+                    : GenerateAuthToken(
+                        "get",
+                        "pkranges",
+                        $"dbs/{_dataStoreConfiguration.DatabaseId}/colls/{_collectionConfiguration.CollectionId}",
+                        date,
+                        key));
+
+            using var httpRequestMessage = new HttpRequestMessage(
+                HttpMethod.Get,
+                $"{host}/dbs/{_dataStoreConfiguration.DatabaseId}/colls/{_collectionConfiguration.CollectionId}/pkranges")
```

Comment:
Interesting, in the v2 sdk it was https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.documents.client.documentclient.readpartitionkeyrangefeedasync?view=azure-dotnet

Reply:
Would have made things easier...